### ISSUE TITLE
[clang-tidy] bugprone-branch-clone false positive with GCCAsmStmt

### ISSUE THREAD

**Issue by hdelassus** [llvm/llvm-project#198616]
When the output and input operands of different inline assembly statements are the same in both branches of an if-else statement, clang-tidy does not know whether the templates are effectively cloned because the AST doesn't parse the assembly template.

Here's a small reproducer for RISC-V with clang-tidy 22.1.0:

```c
// file: foo.c
int main(int argc, char**) {
  if (argc > 1) {
    __asm__ volatile(
      "addi %0, %0, -2"
      : "+r" (argc)
      );
  } else {
    __asm__ volatile(
      "addi %0, %0, -1" // Different immediate
      : "+r" (argc)
      );
  }
  return argc;
}
```

Compile with:
```bash
clang-tidy --config-file=/dev/null --checks=bugprone-branch-clone --extra-arg=-std=gnu23 foo.c
```

AST dump:
```
`-FunctionDecl 0x9ed78e8 <foo.c:1:1, line:14:1> line:1:5 main 'int (int, char **)'
  |-ParmVarDecl 0x9ed7750 <col:10, col:14> col:14 used argc 'int'
  |-ParmVarDecl 0x9ed7808 <col:20, col:25> col:26 'char **'
  `-CompoundStmt 0x9ed7d20 <col:28, line:14:1>
    |-IfStmt 0x9ed7ca8 <line:2:3, line:12:3> has_else
    | |-BinaryOperator 0x9ed7a40 <line:2:7, col:14> 'int' '>'
    | | |-ImplicitCastExpr 0x9ed7a28 <col:7> 'int' <LValueToRValue>
    | | | `-DeclRefExpr 0x9ed79e8 <col:7> 'int' lvalue ParmVar 0x9ed7750 'argc' 'int'
    | | `-IntegerLiteral 0x9ed7a08 <col:14> 'int' 1
    | |-CompoundStmt 0x9ed7ba0 <col:17, line:7:3>
    | | `-GCCAsmStmt 0x9ed7b30 <line:3:5, line:6:7>
    | |   `-DeclRefExpr 0x9ed7b10 <line:5:15> 'int' lvalue ParmVar 0x9ed7750 'argc' 'int'
    | `-CompoundStmt 0x9ed7c90 <line:7:10, line:12:3>
    |   `-GCCAsmStmt 0x9ed7c20 <line:8:5, line:11:7>
    |     `-DeclRefExpr 0x9ed7c00 <line:10:15> 'int' lvalue ParmVar 0x9ed7750 'argc' 'int'
    `-ReturnStmt 0x9ed7d10 <line:13:3, col:10>
      `-ImplicitCastExpr 0x9ed7cf8 <col:10> 'int' <LValueToRValue>
        `-DeclRefExpr 0x9ed7cd8 <col:10> 'int' lvalue ParmVar 0x9ed7750 'argc' 'int'```

**Comment by IamYJLee** [comment]
If we have to handle inline asm, I'd like to solve this issue.
https://godbolt.org/z/6YrjG3jE7

**Comment by vbvictor** [comment]
> If we have to handle inline asm, I'd like to solve this issue. https://godbolt.org/z/6YrjG3jE7

I think we can pretend that every `GccAsmStmt` is different to fix false positive. Parsing inline asm is too much imo.

**Comment by IamYJLee** [comment]
> > If we have to handle inline asm, I'd like to solve this issue. https://godbolt.org/z/6YrjG3jE7
> 
> I think we can pretend that every `GccAsmStmt` is different to fix false positive. Parsing inline asm is too much imo.

Thanks for your comment. I also think parsing and comparing inline asm would be too much.
I’ll work on checking whether the every inline asm strings are identical.

**Comment by IamYJLee** [comment]
I'm sorry for the delay in resolving this issue. I've finally found the root cause of why it's occurring.
The false positive occurs because `VisitGCCAsmStmt` currently passes `AsmString` and `clobber` values to `VisitExpr()`. Since `VisitExpr()` only visits child nodes and `StringLiteral` has no children, the actual string bytes are never included in the profile hash. As a result, two asm statements with different strings end up with the same hash.

Replacing `VisitExpr()` with `Visit()` resolves this by letting the visitor reach `VisitStringLiteral`, which correctly hashes the string bytes. With this change, the false positive no longer occurs.
https://github.com/llvm/llvm-project/blob/db5d4d8e1e79a7cda990d7bb96881eb6316a3055/clang/lib/AST/StmtProfile.cpp#L332

https://github.com/llvm/llvm-project/blob/db5d4d8e1e79a7cda990d7bb96881eb6316a3055/clang/lib/AST/StmtProfile.cpp#L345
